k2/csrc/weights.h

```diff
@@ -29,53 +34,52 @@ namespace k2 {
+   @param [in]  fsa  The fsa we are doing the forward computation on.
+                Must satisfy IsValid(fsa) and IsTopSorted(fsa).
+   @param [in]  arc_weights  Arc weights, indexed by arc in `fsa`.
+                             Usually logprobs.
    @param [out] state_weights  The per-state weights will be written to here.
                 They will be 0 for the start-state (if fsa is
                 nonempty), and for later states will be the
                 largest (most positive) weight from the start-state
-                to that state along any path, or +infinity if no such
+                to that state along any path, or `kNegativeInfinity` if no such
                 path exists.
  */
-void ComputeForwardMaxWeights(const Fsa &fsa, float *state_weights);
+void ComputeForwardMaxWeights(const Fsa &fsa, const float *arc_weights,
+                              float *state_weights);
 
 /*
   Does the 'backward' computation; this is as in the tropical semiring
   but with the opposite sign, as in logprobs rather than negative logprobs.
   It's like shortest path but with the opposite sign.
 
-   @param [in] fsa  The fsa we are doing the backward computation on.
+   @param [in]  fsa  The fsa we are doing the backward computation on.
                 Must satisfy IsValid(fsa) and IsTopSorted(fsa).
+   @param [in]  arc_weights  Arc weights, indexed by arc in `fsa`.
+                             Usually logprobs.
    @param [out] state_weights  The per-state weights will be written to here.
                 They will be 0 for the final-state (if fsa is
                 nonempty), and for earlier states will be the
                 largest (most positive) weight from that
-                to the final state along any path, or +infinity if no such
-                path exists.
+                to the final state along any path, or `kNegativeInfinity` if no
+  such path exists.
  */
-void ComputeBackwardMaxWeights(const Fsa &fsa, float *state_weights);
+void ComputeBackwardMaxWeights(const Fsa &fsa, const float *arc_weights,
+                               float *state_weights);
 
 enum FbWeightType { kMaxWeight, kLogSumWeight };
 
 struct WfsaWithFbWeights {
-  const Fsa *fsa;
+  const Fsa &fsa;
   const float *arc_weights;
-  // forward_state_weights are the log-sum or max of weights along all paths from the
-  // start-state to each state.  We use double because for long FSAs roundoff
-  // effects can cause nasty errors in pruning.
-  const double *forward_state_weights;
-  // backward_state_weights are the log-sum or max of weights along all paths from
-  //  each state to the final state.
-  const double *backward_state_weights;
 
   // Records whether we use max or log-sum.
   FbWeightType weight_type;
 
   /*
     Constructor.
-       @param [in] fsa  Pointer to an FSA; must satisfy
-            IsValid(*fsa) and IsTopSorted(*fsa).
+       @param [in] fsa  Reference to an FSA; must satisfy
```

Comment:
> must satisfy IsValid(fsa) and IsTopSorted(fsa).

But `IsTopSorted(fsa)` is not enforced in the constructor.

Reply:
We will check this in debug mode and in the arc-loops as in `ComputeForwardMaxWeights`.

Added a TODO for now.